src/System.Windows.Extensions/src/System/Media/SoundPlayer.cs

```diff
@@ -0,0 +1,738 @@
+            _currentPos = 0;
+            IsLoadCompleted = false;
+            if (stream != null)
+            {
+                _uri = null;
+            }
+        }
+
+        public void Stop()
+        {
+            Interop.WinMM.PlaySound((byte[])null, IntPtr.Zero, Interop.WinMM.SND_PURGE);
+        }
+
+        public event AsyncCompletedEventHandler LoadCompleted
+        {
+            add
+            {
+                Events.AddHandler(s_eventLoadCompleted, value);
+            }
+            remove
+            {
+                Events.RemoveHandler(s_eventLoadCompleted, value);
+            }
+        }
+
+        public event EventHandler SoundLocationChanged
+        {
+            add
+            {
+                Events.AddHandler(s_eventSoundLocationChanged, value);
+            }
+            remove
+            {
+                Events.RemoveHandler(s_eventSoundLocationChanged, value);
+            }
+        }
+
+        public event EventHandler StreamChanged
+        {
+            add
+            {
+                Events.AddHandler(s_eventStreamChanged, value);
+            }
+            remove
+            {
+                Events.RemoveHandler(s_eventStreamChanged, value);
+            }
+        }
+
+        protected virtual void OnLoadCompleted(AsyncCompletedEventArgs e)
+        {
+            ((AsyncCompletedEventHandler)Events[s_eventLoadCompleted])?.Invoke(this, e);
+        }
+
+        protected virtual void OnSoundLocationChanged(EventArgs e)
+        {
+            ((EventHandler)Events[s_eventSoundLocationChanged])?.Invoke(this, e);
+        }
+
+        protected virtual void OnStreamChanged(EventArgs e)
+        {
+            ((EventHandler)Events[s_eventStreamChanged])?.Invoke(this, e);
+        }
+
+        private void WorkerThread()
+        {
+            try
+            {
+                // setup the http stream
+                if (_uri != null && !_uri.IsFile && _stream == null)
+                {
+                    WebRequest webRequest = WebRequest.Create(_uri);
+
+                    WebResponse webResponse = webRequest.GetResponse();
+
+                    _stream = webResponse.GetResponseStream();
+                }
+
+                _streamData = new byte[BlockSize];
+
+                int readBytes = _stream.Read(_streamData, _currentPos, BlockSize);
+                int totalBytes = readBytes;
+
+                while (readBytes > 0)
+                {
+                    _currentPos += readBytes;
+                    if (_streamData.Length < _currentPos + BlockSize)
+                    {
+                        byte[] newData = new byte[_streamData.Length * 2];
+                        Array.Copy(_streamData, newData, _streamData.Length);
+                        _streamData = newData;
+                    }
+                    readBytes = _stream.Read(_streamData, _currentPos, BlockSize);
+                    totalBytes += readBytes;
+                }
```

Comment:
This whole thread-based mechanism would be better if it were done with async/await, with CancellationToken, etc., replacing that thread abort mechanism, avoiding blocking a thread pool thread, and so on.  Since we need to address the Thread.Abort thing anyway, we may as well clean all this up.

Reply:
(I'm not suggesting we block this PR for it, but I don't think we can ship like this.)